[Pages] Update Nuxt framework guide and improve bindings coverage (#12953)
cloudflare/cloudflare-docs: content/pages/framework-guides/deploy-a-nuxt-site.md

When creating your new project, C3 will give you the option of deploying an initial version of your application via [Direct Upload](/pages/how-to/use-direct-upload-with-continuous-integration/). You can redeploy your application at any time by running following command inside your project directory:

```sh
$ npm run deploy
```

{{<Aside type="note" header="Git integration">}}

The initial deployment created via C3 is referred to as a [Direct Upload](/pages/get-started/direct-upload/). To set up a deployment via the Pages Git integration, refer to the [Git Integration](#git-integration) section below.

{{</Aside>}}

## Configure and deploy a project without C3

To deploy a Nuxt project without C3, follow the [Nuxt Get Started guide](https://nuxt.com/docs/getting-started/installation). After you have set up your Nuxt project, choose either the [Git integration guide](/pages/get-started/guide/) or [Direct Upload guide](/pages/get-started/direct-upload/) to deploy your Nuxt project on Cloudflare Pages.

{{<render file="/_framework-guides/_git-integration.md">}}

### Create a GitHub repository

{{<render file="/_framework-guides/_create-gh-repo.md">}}

```sh

> # Skip the following 3 commands if you have built your application

Reviewer: ```suggestion
# Skip the following three commands if you have built your application
```

# using C3 or already committed your changes
$ git init
$ git add .
$ git commit -m "Initial commit"

$ git branch -M main

> $ git remote add origin https://github.com/<YOUR_GH_USERNAME>/<REPOSITORY_NAME>

Reviewer: ```suggestion
$ git remote add origin https://github.com/<YOUR_GH_USERNAME>/<REPOSITORY_NAME>.git
```

{{<render file="/_framework-guides/_bindings_definition.md">}}

### Set up bindings for local development

Projects created via C3 come with `nitro-cloudflare-dev`, a `nitro` module that simplifies the process of working with bindings during development:

```typescript
---
filename: nuxt.config.ts
highlights: [2]
---
export default defineNuxtConfig({
  modules: ["nitro-cloudflare-dev"],
});
```

This module is powered by the [`getPlatformProxy` helper function](/workers/wrangler/api#getplatformproxy). `getPlatformProxy` will automatically detect any bindings defined in your project's `wrangler.toml` file and emulate those bindings in local development. Review [Wrangler configuration information on bindings](/workers/wrangler/configuration/#bindings) for more information on how to configure bindings in `wrangler.toml`.

{{<Aside type="note">}}

`wrangler.toml` is currently **only** used for local development. Bindings specified in it are not available remotely.

{{</Aside>}}

### Set up bindings for a deployed application

In order to access bindings in a deployed application, you will need to [configure your bindings](/pages/functions/bindings/) in the Cloudflare dashboard.

> ### Add bindings to Typescript projects

Reviewer: ```suggestion
### Add bindings to TypeScript projects
```

To get proper type support, you need to create a new `env.d.ts` file in the root of your project and declare a [binding](/pages/functions/bindings/).

The following is an example of adding a `KVNamespace` binding:

```ts
---
filename: env.d.ts

> highlight: [10]

Reviewer: ```suggestion
highlight: [9]
```

## Create a new project

The easiest way to get started is by using the [`create-cloudflare`](https://www.npmjs.com/package/create-cloudflare) CLI (also known as C3). To do so, open a terminal and run:

```sh
$ npm create cloudflare@latest my-nuxt-app -- --framework=nuxt
```

C3 will create a new project with `nuxi` (The official Nuxt cli) and install the necessary adapters along with the [Wrangler CLI](/workers/wrangler/install-and-update/#check-your-wrangler-version).

> ## Migrating an existing Nuxt application

Reviewer 2: nextjs section needed as well

Add `wrangler` to your project:

```sh
$ npm install --save-dev wrangler@latest
```

Update the deploy script in `package.json`:

```json
---
filename: package.json
---
{
  // ...
  "scripts": {
    // ...
    "deploy": "npm run build && wrangler pages deploy ./dist"
  }
}
```

## Deploy your project

After creating your new project, C3 will give you the option of deploying an initial version of your application via [Direct Upload](/pages/how-to/use-direct-upload-with-continuous-integration/). You can re-deploy your application at any time by running following command inside your project directory:

```sh
$ npm run deploy
```

> You can instead [connect a GitHub or Gitlab repository](/pages/configuration/git-integration) to your Pages project so that new versions of your project are built and deployed when changes to your git repository are detected. To do so, choose “No” when C3 asks if you’d like to deploy and refer to the guide below.

Reviewer 3: In the Next.js guide we went quite a different direction in regards to the git integration: https://github.com/cloudflare/cloudflare-docs/blob/cb0ced81b99626a68d4efe76ff9346e16f03e660/content/pages/framework-guides/nextjs/deploy-a-nextjs-site.md?plain=1#L40-L46

I think we should be consistent and use the Next.js version (which is the one the team agreed on I think)

Reviewer 3: ok I can see the same sort of note below... but I think that this version is more confusing then the Next.js one since here we say "instead" and tell people to say "No" to the C3 deployment, and only later we mention the option to create a new Pages project. 🤔

Reviewer 2: Task for me but we need to make these into partials for these guides. Otherwise, risk of user confusion is too high.

Author: My concern with the Next.js version was that having so much space between explaining the different deployment types seemed confusing to me. This PR reorders things so that both deployment methods are talked about together, one after the other, before moving on to bindings.

I wanted to make sure it's clear to users at the time they're running C3 for the first time that there are 2 deployment pathways: 
- direct upload which you can do by saying "Yes" in c3 and re-deploying via `npm run deploy`, and
- git integration, in which case you should choose "No" when asked to deploy

To me, calling these out and then explaining how to do each in one place is more effective than just having a side note that links to a section later on down the page.

Reviewer 3: The Next.js version was written after our team discussion and my understanding from that was that the intention was to always suggest people to deploy via C3 (say "Yes") and then adding the direct upload option in the guide as an aside (e.g. by the way, you can also do git-integration if you want).

The argument being that we don't want to put too many choices upfront to the user, making the guide less straightforward to follow.

(Also when after adding git-integration support to already deployed Pages application is implemented the above structure would be quite easy to update and already well structured for the new workflow)

Then we agreed on following Next.js' example on this for the other guides.

I am not saying that I necessarily prefer one version over the other, but if you have concerns about the Next.js version I think it might be worth to re-discuss it with the team, agree (or not) on the different direction and then apply it to all guides moving forward (updating Next.js as well)?

(PS: my only concern here is that for each guide, who updates it can put their own opinion based spin on this issue without an agreed consistent way of presenting this to the user, while I think it'd be much better to clearly define what we want to go with and just consistently rinse and repeat that)

$ git push -u origin main
```

### Create a Pages project

1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/) and select your account.
2. Navigate to [Workers & Pages > Create application > Pages > Connect to Git](https://dash.cloudflare.com/?to=/:account/pages/new/provider/github) and create a new pages project.

{{<Aside type="note">}}

Note that git integration cannot currently be added to existing Pages applications, so if you've already deployed your application via Direct Upload (using C3 for example), you will still need to create a new Pages application in order to proceed.

{{</Aside>}}

You will be asked to authorize access to your GitHub account if you have not already done so. Cloudflare needs this so that it can monitor and deploy your projects from the source. You may narrow access to specific repositories if you prefer; however, you will have to manually update this list [within your GitHub settings](https://github.com/settings/installations) when you want to add more repositories to Cloudflare Pages.

Select the new GitHub repository that you created and, in the **Set up builds and deployments** section, provide the following information:

> {{<pages-build-preset framework="nuxt">}}

Reviewer 3: I don't think this change is correct:
![Screenshot 2024-02-13 at 18 41 10](https://github.com/cloudflare/cloudflare-docs/assets/61631103/e9e1cee8-07a8-4604-9d9a-971929ecae37)

If you want to rename the preset you'll also have to rename it here I think:
https://github.com/cloudflare/cloudflare-docs/blob/cb0ced81b99626a68d4efe76ff9346e16f03e660/content/pages/_partials/_build-configuration.md?plain=1#L28

But I'd be careful about the above since I think this file is also used to generate the dashboard presets... so before updating it I double check that it is ok to do so

{{</Aside>}}

To get proper type support, you need to create a new `env.d.ts` file in your project and declare a [binding](/pages/functions/bindings/).

The following is an example of adding a `KVNamespace` binding:

```ts
---
filename: env.d.ts
highlight: [10]
---
import { CfProperties, Request, ExecutionContext, KVNamespace } from '@cloudflare/workers-types';

declare module 'h3' {
    interface H3EventContext {
        waitUntil: ExecutionContext['waitUntil']
        cf: CfProperties,
        cloudflare: {
          request: Request,
          env: {
            MY_KV: KVNamespace,
          }
          context: ExecutionContext,
        };
    }
}
```

### Access bindings in your Nuxt application

In Nuxt, add server-side code via [Server Routes and Middleware](https://nuxt.com/docs/guide/directory-structure/server#server-directory). The `defineEventHandler()` method is used to define your API endpoints in which you can access Cloudflare's context via the provided `context` field. The `context` field allows you to access any bindings set for your application.

The following code block shows an example of accessing a KV namespace in Nuxt.

```typescript
---

> filename: server/api/hello.ts / server/api/hello.js

Reviewer 3: @deadlypants1973 this is _very_ ugly! 😓 

![Screenshot 2024-02-22 at 01 27 12](https://github.com/cloudflare/cloudflare-docs/assets/61631103/a7d38aca-4831-4018-a8bc-2f1fd54fe8b0)

I'm always unsure what's best in these situations... shall we add a js/ts switcher even if completely unnecessary? (since the file content is always the same)

Reviewer 2: +1 for switcher!

Reviewer 3: Switcher added 🙂👍 

Although it showcases a small coloring quirk 😓 
![switcher](https://github.com/cloudflare/cloudflare-docs/assets/61631103/3326af2e-6548-488e-97b9-8531159383fa)

Reviewer 3: (PS: the coloring difference seems a bit wrong to me... I wonder if it is intentional...)

Additionally, you will have access to [preview deployments](/pages/configuration/preview-deployments/), which repeat the build-and-deploy process for pull requests. With these, you can preview changes to your project with a real URL before deploying them to production.

## Bindings

A [binding](/pages/functions/bindings/) allows your application to interact with Cloudflare developer products, such as [KV](/kv/reference/how-kv-works/), [Durable Object](/durable-objects/), [R2](/r2/), and [D1](https://blog.cloudflare.com/introducing-d1/).

If you intend to use bindings in your project, you must first set them up for local and remote development.

Projects created via C3 come with `nitro-cloudflare-dev`, a `nitro` module that simplifies the process of working with bindings during development:

```typescript
---
filename: nuxt.config.ts
---
export default defineNuxtConfig({
  modules: ["nitro-cloudflare-dev"],
});
```

> This module is powered by the `getBindingsProxy` [helper function](/workers/wrangler/api#getbindingsproxy) which will automatically detect any bindings defined in the `wrangler.toml` file and emulate them in local development. See [Wrangler configuration](workers/wrangler/configuration/#bindings) for more information on how to configure bindings in `wrangler.toml`.

Reviewer 3: This section will be outdates after the changes in https://github.com/cloudflare/cloudflare-docs/pull/12957 are applied

I don't think it should be updated now, but if #12957 happens to be merged before this PR then this section should be updated (otherwise we can update it in #12957 or in a followup PR)

I just figured I'd mention it 🙂

> This module is powered by the `getBindingsProxy` [helper function](/workers/wrangler/api#getbindingsproxy) which will automatically detect any bindings defined in the `wrangler.toml` file and emulate them in local development. See [Wrangler configuration](workers/wrangler/configuration/#bindings) for more information on how to configure bindings in `wrangler.toml`.

Reviewer 3: since https://github.com/cloudflare/workers-sdk/pull/5010 got merged:
```suggestion
This module is powered by the `getPlatformProxy` [helper function](/workers/wrangler/api#getplatformproxy) will automatically detect any bindings defined in the `wrangler.toml` file and emulate them in local development. See [Wrangler configuration](workers/wrangler/configuration/#bindings) for more information on how to configure bindings in `wrangler.toml`.
```

{{<Aside type="note">}}
`wrangler.toml` is currently **only** used for local development. Bindings specified in it are not available remotely.
{{</Aside>}}

In order to access bindings in a deployed application, you will need to [configure them](/pages/functions/bindings/) in the Cloudflare dashboard on your project's settings page.

> ### Accessing bound resources in your Nuxt application

Reviewer 3: calling "bindings" "bound resources" make things more confusing in my opinion 😅
